rfcs/0110-inherit-as-list.md

```diff
@@ -0,0 +1,153 @@
+---
+feature: inherit-as-list
+start-date: 2021-10-17
+author: Ryan Burns (@r-burns)
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+This RFC proposes a new Nix syntax `[ inherit (<attrset>) <attrnames>; ]`,
+which constructs a list from the values of an attrset.
+
+The goal is to provide a similarly-terse but more principled alternative
+to the often-used `with <attrset>; [ <attrnames> ]`.
+
+# Motivation
+[motivation]: #motivation
+
+It is currently cumbersome to create a list from the values of an attrset.
+If one has an attrset `attrs` and wishes to create a list containing some of
+its values, one could naively write:
+
+```nix
+[ attrs.a attrs.b attrs.c ]
+```
+
+To avoid typing `attrs` many times, one will typically use `with` instead:
+
+```nix
+with attrs; [ a b c ]
+```
+
+However, the `with` expression has many well-known drawbacks, such as
+unintuitive shadowing behavior [1][2], prevention of static scope checking [3][4],
+and reduced evaluation performance [3].
+
+* [1] https://github.com/NixOS/nix/issues/490
+* [2] https://github.com/NixOS/nix/issues/1361
+* [3] https://github.com/NixOS/nixpkgs/pull/101139
+* [4] https://nix.dev/anti-patterns/language#with-attrset-expression
+
+Nonetheless, Nix expression authors are subtly guided toward the `with` form
+because it is (or at least appears) simpler than any existing alternatives.
+Some alternatives are suggested in
+https://nix.dev/anti-patterns/language#with-attrset-expression, but as these
+are more verbose and complex than `with`, they are rarely used.
+
+The goal of this RFC is to provide a similarly-terse alternative which avoids
+these drawbacks.
+
+# Detailed design
+[design]: #detailed-design
+
+The proposed syntax is:
+
+```
+[ inherit (attrs) a b c; ]
+[ inherit a b c; ]
```

Comment:
Isn't this equivalent to just `[a b c]`. Is it allowed just for consistency?

Reply:
If ` [ inherit a b c; ]` is equivalent to `[ a b c ]`, I'm still slightly inclined to allow it for the sake of consistency, but I don't really have a strong opinion on it. What do you think?

Reply:
This is probably a personal preference between consistency and redundancy. Personally I would prefer avoiding too much redundancy. Redundancy means that when I am reading Nix expressions I have to normalize the two synatxes in my head. I also take a moment to try and figure out why it was done this way rather than the way I consider more obvious. A formatter could probably remove the inherit but that requires additional tooling with its own downsides. 

Basically I don't want to see `[inherit a b c]` and wonder what it means. I have enough Nix syntax in my head and want to avoid any extra, even if it is similar to the object inherit I still need to remember how it works and that it is in fact completely unnecessary noise on the line.